### PR TITLE
Enchanted link

### PR DESCRIPTION
**Pull request by aviadl**
+ tests
Remove support from cross device
related to https://github.com/descope/etc/issues/1014

### CODE REVIEW

File: examples/webapp/main.go [hunk 1]
```diff
@@ -135,6 +140,9 @@ func help(w http.ResponseWriter, r *http.Request) {
 	helpTxt += "Sign in of existing user with magiclink sms go to: /magiclink/signin?sms=\n\n"
 	helpTxt += "Sign in of existing user with magiclink whatsapp go to: /magiclink/signin?whatsapp=\n\n"
 	helpTxt += "---------------------------------------------------------\n\n"
+	helpTxt += "Sign up with enchantedlink go to: /enchantedlink/signup?email=\n\n"
```
Comment (dorsha) [inline examples/webapp/main.go:143]: Change to "enchanted link", when they are connected it is hard to read.

File: examples/webapp/main.go [hunk 2]
```diff
@@ -260,6 +268,73 @@ func handleMagicLinkVerify(w http.ResponseWriter, r *http.Request) {
...
 }
 
+func handleEnchantedLinkSignIn(w http.ResponseWriter, r *http.Request) {
+	_, identifier := getMethodAndIdentifier(r)
+	enchantedRes, err := client.Auth.EnchantedLink().SignIn(identifier, verifyMagicLinkURI, nil, nil)
+	if err != nil {
+		setErrorWithSignUpIn(w, err.Error(), "", identifier)
+		return
+	}
+	helpTxt := "You should have received an enchantedlink by email\n"
```
Comment (dorsha) [inline examples/webapp/main.go:278]: Same (do this in all places)

File: descope/auth/enchantedlink.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+	if identifier == "" {
+		return nil, errors.NewInvalidArgumentError("identifier")
+	}
+	if loginOptions.IsJWTRequired() {
+		pswd, err = getValidRefreshToken(r)
+		if err != nil {
+			return nil, errors.InvalidStepupJwtError
+		}
+	}
+	httpResponse, err := auth.client.DoPostRequest(composeEnchantedLinkSignInURL(), newMagicLinkAuthenticationRequestBody(identifier, URI, true, loginOptions), nil, pswd)
```
Comment (dorsha) [inline descope/auth/enchantedlink.go:25]: Shouldn't it be newEnchantedLinkAuthenticationRequestBody?

File: descope/auth/enchantedlink.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+		return nil, errors.NewInvalidArgumentError("identifier")
+	}
+	if user == nil {
+		user = &User{}
+	}
+	if len(user.Email) == 0 {
+		user.Email = identifier
+	}
+
+	httpResponse, err := auth.client.DoPostRequest(composeEnchantedLinkSignUpURL(), newMagicLinkAuthenticationSignUpRequestBody(MethodEmail, identifier, URI, user, true), nil, "")
```
Comment (dorsha) [inline descope/auth/enchantedlink.go:43]: same

File: descope/auth/enchantedlink.go [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+		return nil, err
+	}
+	return getPendingRefFromResponse(httpResponse)
+}
+
+func (auth *enchantedLink) SignUpOrIn(identifier, URI string) (*EnchantedLinkResponse, error) {
+	if identifier == "" {
+		return nil, errors.NewInvalidArgumentError("identifier")
+	}
+	httpResponse, err := auth.client.DoPostRequest(composeEnchantedLinkSignUpOrInURL(), newMagicLinkAuthenticationRequestBody(identifier, URI, true, nil), nil, "")
```
Comment (dorsha) [inline descope/auth/enchantedlink.go:54]: same

File: descope/auth/enchantedlink.go [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+	}
+	return getPendingRefFromResponse(httpResponse)
+}
+
+func (auth *enchantedLink) GetSession(pendingRef string, w http.ResponseWriter) (*AuthenticationInfo, error) {
+	var err error
+	httpResponse, err := auth.client.DoPostRequest(composeGetSession(), newAuthenticationGetMagicLinkSessionBody(pendingRef), nil, "")
+	if err != nil {
+		if err == errors.UnauthorizedError {
+			return nil, errors.MagicLinkUnauthorized
```
Comment (dorsha) [inline descope/auth/enchantedlink.go:66]: Shouldn't it be errors.EnchantedLinkUnauthorized?

File: descope/auth/enchantedlink.go [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+		}
+		return nil, err
+	}
+	return auth.generateAuthenticationInfo(httpResponse, w)
+}
+
+func (auth *enchantedLink) Verify(token string) error {
+	_, err := auth.client.DoPostRequest(composeVerifyEnchantedLinkURL(), newMagicLinkAuthenticationVerifyRequestBody(token), nil, "")
```
Comment (dorsha) [inline descope/auth/enchantedlink.go:74]: same

File: descope/auth/enchantedlink.go [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+		return nil, errors.NewInvalidArgumentError("email")
+	}
+	if !emailRegex.MatchString(email) {
+		return nil, errors.NewInvalidArgumentError("email")
+	}
+	pswd, err := getValidRefreshToken(r)
+	if err != nil {
+		return nil, err
+	}
+	httpResponse, err := auth.client.DoPostRequest(composeUpdateUserEmailEnchantedLink(), newMagicLinkUpdateEmailRequestBody(identifier, email, URI, true), nil, pswd)
```
Comment (dorsha) [inline descope/auth/enchantedlink.go:95]: same

File: descope/auth/enchantedlink_test.go [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+func TestSignInEnchantedLinkStepupNoJwt(t *testing.T) {
+	email := "test@test.com"
+	a, err := newTestAuth(nil, nil)
+	require.NoError(t, err)
+	_, err = a.EnchantedLink().SignIn(email, "", nil, &LoginOptions{Stepup: true})
+	require.Error(t, err)
+	assert.ErrorIs(t, err, errors.InvalidStepupJwtError)
+}
+
+//			assert.EqualValues(t, composeMagicLinkSignInURL(MethodEmail), r.URL.RequestURI())
```
[the remark this answers is not in the document]
Comment (aviadl) [reply descope/auth/enchantedlink_test.go]: Mistake, removed

File: examples/webapp/main.go (hunk 1 above)
Comment (aviadl) [reply examples/webapp/main.go]: done

File: examples/webapp/main.go (hunk 2 above)
Comment (aviadl) [reply examples/webapp/main.go]: done

File: descope/auth/enchantedlink.go (hunk 3 above)
Comment (aviadl) [reply descope/auth/enchantedlink.go:25]: Left as same body, since those are the same params

File: descope/auth/enchantedlink.go (hunk 5 above)
Comment (aviadl) [reply descope/auth/enchantedlink.go:54]: .

File: descope/auth/enchantedlink.go (hunk 6 above)
Comment (aviadl) [reply descope/auth/enchantedlink.go:66]: fixed

File: descope/auth/enchantedlink.go (hunk 7 above)
Comment (aviadl) [reply descope/auth/enchantedlink.go:74]: .

File: descope/auth/enchantedlink.go (hunk 8 above)
Comment (aviadl) [reply descope/auth/enchantedlink.go:95]: .

**Review by dorsha — APPROVED** [review]

**Pull request merged**